src/codegen/exeffToNoeff.ml

```diff
@@ -0,0 +1,360 @@
+  match effects with
+  | (((eff, _), abs)::es) -> get_effectset (EffectSet.add eff set) es
+  | [] -> set
+
+and subst_ty_param tysub par ty =
+  match ty with
+  | ExEffTypes.TyParam x -> if (x = par) then (tysub) else (ty)
+  | ExEffTypes.Apply (n, ls) -> ExEffTypes.Apply (n, List.map (subst_ty_param tysub par) ls)
+  | ExEffTypes.Arrow (l, (rt, rd)) -> ExEffTypes.Arrow (subst_ty_param tysub par l, (subst_ty_param tysub par rt, rd))
+  | ExEffTypes.Tuple ls -> ExEffTypes.Tuple (List.map (subst_ty_param tysub par) ls)
+  | ExEffTypes.Handler ((lt, ld), (rt, rd)) -> ExEffTypes.Handler ((subst_ty_param tysub par lt, ld), (subst_ty_param tysub par rt, rd))
+  | ExEffTypes.PrimTy p -> ExEffTypes.PrimTy p
+  | ExEffTypes.QualTy ((ty1, ty2), ty3) ->
+    ExEffTypes.QualTy ((subst_ty_param tysub par ty1, subst_ty_param tysub par ty2), 
+        subst_ty_param tysub par ty3)
+  | ExEffTypes.QualDirt (dirts, t) -> ExEffTypes.QualDirt (dirts, subst_ty_param tysub par t)
+  | ExEffTypes.TySchemeTy (p, skel, t) -> ExEffTypes.TySchemeTy (p, skel, subst_ty_param tysub par t)
+  | ExEffTypes.TySchemeDirt (p, t) -> ExEffTypes.TySchemeDirt (p, subst_ty_param tysub par t)
+  | ExEffTypes.TySchemeSkel (p, t) -> ExEffTypes.TySchemeSkel (p, subst_ty_param tysub par t)
+ 
+and comp_elab state env c = 
+  match c with 
+  | ExEff.Value value ->
+    let (t, elab) = value_elab state env value in
+    ((t, ExEffTypes.empty_dirt), elab)
+  | ExEff.LetVal (value, (pat, _, comp)) -> 
+    let (tyv, elabv) = value_elab state env value in
+    let env' = extend_pattern_type env pat tyv in
+    let (tyc, elabc) = comp_elab state env' comp in
+    (tyc, NoEff.NLet (elabv, (pattern_elab pat, elabc)))
+  | ExEff.LetRec (abs_list, comp) ->
+    let rec extend_env env ls = 
+      ( match ls with
+        | [] -> env
+        | ( (var, ty1, ty2, (p, comp)) :: rest ) ->
+        let env' = TypeChecker.extend_var_types env var (ExEffTypes.Arrow (ty1, ty2)) in
+        let env'' = extend_pattern_type env' p ty1 in
+        extend_env env'' rest ) in
+      let elab_letrec_abs ( (var, ty1, ty2, (p, compt)) ) =
+              ( let (_, t1) = type_elab state env ty1 in
+              let (_, t2) = dirty_elab state env ty2 in
+              let (_, elabc) = comp_elab state env compt in
+              ( (var, t1, t2, (pattern_elab p, elabc)) ) ) in
+      let (tycomp, elabcomp) = comp_elab state (extend_env env abs_list) comp in
+      ( tycomp, NoEff.NLetRec (List.map (elab_letrec_abs) abs_list, elabcomp) )
+  | ExEff.Match (value, abs_lst, loc) ->
+    let (tyv, elabv) = value_elab state env value in
+    let elab_abs (pat, comp) = 
+            ( let (tyc, elabc) = comp_elab state env comp in (pattern_elab pat, elabc) ) in
+    ( if ( (List.length abs_lst) = 0)
+    then ( failwith "Empty match statement" )
+    else ( let ((p1,c1) :: _ ) = abs_lst in
+           let (tyc, elabc) = comp_elab state env c1 in 
+           (tyc, NoEff.NMatch (elabv, List.map elab_abs abs_lst, loc)) ) )
```

Comment:
You forgot to check a couple of things here:
* **ALL** left-hand sides (patterns) must have the same type (in this instance `tyv`)
* **ALL** right-hand sides (computations) must have the same type (in this instance `tyc`)
You need to check these explicitly.